### ISSUE TITLE
CCpp target does not initialize C++ classes that are used as state variables properly

### ISSUE THREAD

**Issue by cmnrd** [lf-lang/lingua-franca#741]
This was causing the segfault in #712. Here is a minimal program to reproduce the segfault:
```
target CCpp;

preamble {=
    #include <map>
=}

main reactor {
    state map: std::map<int, int>;

    reaction(startup) {=
        self->map[0] = 1;
    =}
}
```
The same program works fine for the C++ target (a few small adjustments are needed for the C++ target, but the essential parts are the same). To me this looks as if the default constructor on the `map` state variable is never called. Or it is called, but then something invalidates the internal state of `map` (which would be more worrying). The program above can be fixed by manually constructing a map and assigning it `self->map = std::map<int, int>()`.

Is it required for all state in the CCpp target to be initialized in a startup reaction? If so, why does it seem to work for an `std::vector` but not for a map?

**Comment by Soroosh129** [comment]
I don't think we call the constructor of the state variables at all. The struct for a reactor is "instantiated" using `calloc`, which does not call constructors (like `new` would do). I have attempted to use `new` before but it required changing too many things in the C code generator. I have forgotten the specific issues I ran into though.

> Is it required for all state in the CCpp target to be initialized in a startup reaction?

If a constructor needs to be called, yes.

> If so, why does it seem to work for an std::vector but not for a map?

I'm not sure. If `std::vector` is working, then somehow calling its constructor is not a requirement, or it's happy with the zeroed-out memory that it gets.

**Comment by cmnrd** [comment]
If that is the case, then we should have a big disclaimer somewhere that says all state variables should be initialized manually and the C++ default initialization rules do not apply. I personally think that this behavior of the CCpp target is very dangerous, and it should use a C++ mechanism to construct the self struct.

> I'm not sure. If std::vector is working, then somehow calling its constructor is not a requirement, or it's happy with the zeroed-out memory that it gets.

It seems more that it "works" by chance and nothing broke yet.

**Comment by Soroosh129** [comment]
I think you're right. This behavior is dangerous.
